autofaiss/indices/distributed.py

```diff
@@ -336,3 +342,318 @@ def run(
+    index, metrics = _add_embeddings_from_dir_to_index(
+        add_embeddings_fn=partial(
+            add_embeddings_to_index_distributed,
+            trained_index_or_path=trained_index_path,
+            nb_cores=nb_cores,
+            temporary_indices_folder=partition_temp_root_dir,
+            nb_indices_to_keep=nb_splits_per_big_index,
+        ),
+        embedding_root_dir=embedding_root_dir,
+        output_root_dir=output_root_dir,
+        index_key=trained_index_key,
+        embedding_column_name=embedding_column_name,
+        current_memory_available=current_memory_available,
+        id_columns=id_columns,
+        max_index_query_time_ms=max_index_query_time_ms,
+        min_nearest_neighbors_to_retrieve=min_nearest_neighbors_to_retrieve,
+        use_gpu=use_gpu,
+        make_direct_map=make_direct_map,
+    )
+
+    # Only metrics are returned to save memory on driver
+    if index:
+        del index
+
+    return metrics
+
+
+def create_small_index(
+    embedding_root_dir: str,
+    output_root_dir: str,
+    id_columns: Optional[List[str]] = None,
+    should_be_memory_mappable: bool = False,
+    max_index_query_time_ms: float = 10.0,
+    max_index_memory_usage: str = "16G",
+    min_nearest_neighbors_to_retrieve: int = 20,
+    embedding_column_name: str = "embedding",
+    current_memory_available: str = "32G",
+    use_gpu: bool = False,
+    metric_type: str = "ip",
+    nb_cores: Optional[int] = None,
+    make_direct_map: bool = False,
+) -> Tuple[Optional[faiss.Index], Optional[Dict[str, str]]]:
+    """
+    Create a small index
+    """
+
+    # Train index
+    trained_index = create_and_train_index_from_embedding_dir(
+        embedding_root_dir=embedding_root_dir,
+        embedding_column_name=embedding_column_name,
+        max_index_memory_usage=max_index_memory_usage,
+        make_direct_map=make_direct_map,
+        should_be_memory_mappable=should_be_memory_mappable,
+        use_gpu=use_gpu,
+        metric_type=metric_type,
+        nb_cores=nb_cores,
+        current_memory_available=current_memory_available,
+        id_columns=id_columns,
+    )
+
+    # Add embeddings to index and compute metrics
+    return _add_embeddings_to_index(
+        add_embeddings_fn=partial(
+            add_embeddings_to_index_local,
+            trained_index_or_path=trained_index.index_or_path,
+            add_embeddings_with_ids=True,
+        ),
+        embedding_reader=trained_index.embedding_reader_or_path,
+        output_root_dir=output_root_dir,
+        index_key=trained_index.index_key,
+        current_memory_available=current_memory_available,
+        id_columns=id_columns,
+        max_index_query_time_ms=max_index_query_time_ms,
+        min_nearest_neighbors_to_retrieve=min_nearest_neighbors_to_retrieve,
+        use_gpu=use_gpu,
+        make_direct_map=make_direct_map,
+    )
+
+
+def create_partitioned_indexes(
```

Comment:
This will train a new index for each partition and store each index trained parameters independently

It works, and is necessary if embeddings space are different. It might improve search a bit even if embeddings spaces are the same.
Drawbacks are:
- training the indices N times
- storing the parameters N times. That is particularly a problem if many small indices 

An alternative in case of same embedding space could be to train the index only once, then save the training parameters only once and save each partition files without the training parameters.
This is particularly possible with the on disk inverted list implementation.

Might be an option in the future, of course not blocking this PR, already pretty nice :)